source/MaterialXGenShader/Util.cpp

```diff
@@ -570,10 +700,128 @@ bool elementRequiresShading(ConstTypedElementPtr element)
+                          std::vector<ParameterPtr>& parameters,
+                          std::vector<OutputPtr>& inputSources,
+                          std::vector<OutputPtr>& parameterSources)
+{
+    if (shaderNode->getType() != SURFACE_SHADER_TYPE_STRING &&
+        shaderNode->getType() != DISPLACEMENT_SHADER_TYPE_STRING &&
+        shaderNode->getType() != VOLUME_SHADER_TYPE_STRING)
+    {
+        return;
+    }
+
+    for (ParameterPtr parameter : shaderNode->getParameters())
+    {
+        OutputPtr output = parameter->getConnectedOutput();
+        if (output)
+        {
+            parameters.push_back(parameter);
+            parameterSources.push_back(output);
+        }
+    }
+
+    for (InputPtr input : shaderNode->getInputs())
+    {
+        OutputPtr output = input->getConnectedOutput();
+        if (output)
+        {
+            inputs.push_back(input);
+            inputSources.push_back(output);
+        }
+    }
+}
 
+vector<MaterialAssignPtr> getGeometryBindings(NodePtr materialNode, const string& geom)
+{
+    vector<MaterialAssignPtr> matAssigns;
+    for (LookPtr look : materialNode->getDocument()->getLooks())
+    {
+        for (MaterialAssignPtr matAssign : look->getMaterialAssigns())
+        {
+            if (matAssign->getReferencedMaterialNode() == materialNode)
+            {
+                if (geomStringsMatch(geom, matAssign->getActiveGeom()))
+                {
+                    matAssigns.push_back(matAssign);
+                    continue;
+                }
+                CollectionPtr coll = matAssign->getCollection();
+                if (coll && coll->matchesGeomString(geom))
+                {
+                    matAssigns.push_back(matAssign);
+                    continue;
+                }
+            }
+        }
+    }
+    return matAssigns;
+}
+
+void findRenderableMaterialNodes(ConstDocumentPtr doc, 
+                                 vector<TypedElementPtr>& elements, 
+                                 bool includeReferencedGraphs,
+                                 std::unordered_set<OutputPtr> &processedOutputs)
+{
+    for (const NodePtr& material : doc->getMaterialNodes())
+    {
+        // Push the material node only once.
+        elements.push_back(material);
+
+        vector<NodePtr> shaderNodes = getShaderNodes(material);
+        for (NodePtr shaderNode : shaderNodes)
+        {
+            if (!includeReferencedGraphs)
+            {
+                // Track outputs already used by the shader
+                std::vector<InputPtr> inputs;
+                std::vector<ParameterPtr> parameters;
+                std::vector<OutputPtr> inputSources;
+                std::vector<OutputPtr> parameterSources;
+                getShaderConnections(shaderNode,
+                                     inputs, parameters,
+                                     inputSources, parameterSources);
+
+                for (size_t i=0; i<inputSources.size(); i++)
```

Comment:
Only inputSources seems to be used below, so inputs, parameters and parameterSources are extracted unnecessarily. Would it not be better to just iterate on the inputs and finding the outputs directly here? Is the function `getShaderConnections` even needed, it seems to only be used here.

Reply:
I can split this up, but to avoid errors it's still good to wrap this up. I actually need to patch an omission on the method. I can do getShaderInputConnections() and getShaderParameterConnections()